File: leetcode/dia 045/totalNQueens.py
class Solution:
    def totalNQueens(self, n: int):
        from copy import deepcopy
        def rec(i, c, dp, ds):
            total = 0
            for j in range(n):
                if not j in c and not (i- min(i, j) , j - min(i, j)) in dp and not (i - min(i, n - 1 - j), j + min(i, n - 1 - j)) in ds:
                    if i == n - 1:
                        total += 1
                    else:
                        c.add(j)
                        dp.add((i- min(i, j) , j - min(i, j)))
                        ds.add((i - min(i, n - 1 - j), j + min(i, n - 1 - j)))
                        total += rec(i + 1, c, dp, ds)
                        c.remove(j)
                        dp.remove((i- min(i, j) , j - min(i, j)))
                        ds.remove((i - min(i, n - 1 - j), j + min(i, n - 1 - j)))
            return total
        return rec(0, set(), set(), set())

File: leetcode/dia 055/totalNQueens.cpp
class Solution {

int t = 0;
vector <bool> c;
vector <bool> d1;
vector <bool> d2;

void search(int i, int n){
    if (i == n){
        t ++;
    } 
    for (int j = 0; j < n; j++){
        if (c[j] || d1[i + j] || d2[i - j + n - 1]) continue;
        c[j] = d1[i + j] = d2[i - j + n - 1] = true;
        search(i + 1, n);
        c[j] = d1[i + j] = d2[i - j + n - 1] = false;
    }
}

public:
    int totalNQueens(int n) {
        c.assign(n, false);
        d1.assign(2 * n - 1, false);
        d2.assign(2 * n - 1, false);
        search(0, n);
        return t;
    }
};